pkg/apis/core/v1/validation/validation.go

```diff
@@ -142,6 +149,14 @@ func ValidatePodLogOptions(opts *v1.PodLogOptions) field.ErrorList {
 			allErrs = append(allErrs, field.Invalid(field.NewPath("sinceSeconds"), *opts.SinceSeconds, "must be greater than 0"))
 		}
 	}
+	if opts.Stream != nil {
```

Comment:
please comment the nil check

Reply:
Re defaulting:

Typically when I think about defaulting I think of it from the client perspective-- If a field is not set in a write, then the client *sees* the default when reading the resource.

But for query parameters, there is nothing for the client to see. The query params are sent in request and that's it.

Query parameters are so rarely specified, that the code to handle it is very manual:.https://github.com/kubernetes/kubernetes/blob/37ed906a33211c7d578cab2d681941ebfd2f2f23/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L118-L127

We don't even consistently use enums: https://github.com/kubernetes/kubernetes/blob/574460e21a9de05c487b1498dc1abbebf97b92eb/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L559

So my take is:  Accept strings or typerefs to strings (since that's what query params are), then make sure your handler does the right validation and marshaling to a struct in the format that you need.  If you can use generated defaulting by calling it by hand, go for it. If not, it's fine to do it manually.

Sorry the api machinery is so undeveloped for query params.

Reply:
In this case, the Decode *works* and even calls defaulting, so consumer code can ALMOST rely on a valid value (except in the case of 0 QPs, which I think we should fix).

Agree that it's relatively niche, and so bending rules is OK.  I think the solution to the openapi problem is just not to use a typeref, just make it a `string` or `*string` field

Reply:
> In this case, the Decode works and even calls defaulting

Excellent, I didn't realize that.

> except in the case of 0 QPs, which I think we should fix

Can we band-aid this and then follow up with a machinery fix post-freeze?

> I think the solution to the openapi problem is just not to use a typeref, just make it a `string` or `*string` field

:+1:

Reply:
> I think the solution to the openapi problem is just not to use a typeref, just make it a string or *string field

Working on it.

Reply:
> please comment the nil check

Updated

Reply:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/127360/pull-kubernetes-integration/1854167462572986368

> F1106 14:27:10.116401   23282 openapi.go:43] Failed to build open api spec for root: non-body Restful parameter type should be a simple type, but got : *string
!!! [1106 14:27:39] Timed out waiting for apiserver to answer at https://127.0.0.1:6443/healthz; tried 30 waiting 1 between each
!!! [1106 14:27:39] Call tree:
!!! [1106 14:27:39]  1: hack/make-rules/test-cmd.sh:93 kube::util::wait_for_url_with_bearer_token(...)
!!! [1106 14:27:39]  2: hack/make-rules/test-cmd.sh:196 run_kube_apiserver(...)

The openapi problem still persists..